### DISCUSSION TITLE
Query Historical Data with Read At Time

### DISCUSSION THREAD

**Discussion by alexschoeberl** [FreeOpcUa/opcua-asyncio#1609] [Q&A]
The OPC UA documentation for historical read of variables specifies different functionalities of the server, including "Read raw", "Read modified" and "Read at time". The "Read at time" option says: "This structure selects a set of raw or interpolated values from the history database by specifying a series of timestamps".

I did not find a method of the client class that implements this functionality. How can I read historical values at given timestamps or at a specified interval?

**Comment by alexschoeberl** [comment]
Any suggestions? Is this functionality simply not provided by the package?

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/8fc26b88c2f5f7730cf911999b94076beeca8272/asyncua/client/ua_client.py#L712

**Reply by AndreasHeine** [reply]
unfortunatly there is no high level method...

**Reply by schroeder-** [reply]
Search for read_raw_history

**Reply by AndreasHeine** [reply]
client.uaclient.history_read(...)

**Reply by AndreasHeine** [reply]
params is:
https://github.com/FreeOpcUa/opcua-asyncio/blob/8fc26b88c2f5f7730cf911999b94076beeca8272/asyncua/ua/uaprotocol_auto.py#L7242

which contains "HistoryReadDetails" which is https://reference.opcfoundation.org/v104/Core/docs/Part11/6.4.1

**Reply by alexschoeberl** [reply]
Thank you for the hint @AndreasHeine! I tried the following:

```
@dataclass(frozen=FROZEN)
class ReadAtTimeDetails(HistoryReadDetails):
    data_type = NodeId(ObjectIds.ReadAtTimeDetails)

    ReqTimes: List[UtcTime] = field(default_factory=list)
    UseSimpleBounds: Boolean = True

node = client.get_node("ns=2;...")
details = ReadAtTimeDetails()
details.ReqTimes = [datetime.utcnow() - timedelta(seconds=170), datetime.utcnow() - timedelta(seconds=130)]
details.UseSimpleBounds = True
history = []
continuation_point = None
while True:
    result = await node.history_read(details, continuation_point)
    result.StatusCode.check()
    continuation_point = result.ContinuationPoint
    history.extend(result.HistoryData.DataValues)  # type: ignore[attr-defined]
    # No more data available
    if continuation_point is None:
            break
```
 
However, I always get this error:
asyncua.ua.uaerrors._auto.BadNotImplemented: Requested operation is not implemented.(BadNotImplemented)

Am I missing something?

**Reply by AndreasHeine** [reply]
BadNotImplemented comes from Server...

does it work with UAExpert?

**Reply by schroeder-** [reply]
Most servers only support RawHistoryAccess. To do this you can copy the code of:
https://github.com/FreeOpcUa/opcua-asyncio/blob/8fc26b88c2f5f7730cf911999b94076beeca8272/asyncua/common/node.py#L652

**Reply by alexschoeberl** [reply]
I see, so most likely the server does not allow it. I wanted to reduce the number of values I need to send in a request, only get one value every 30 seconds.
I don't think there is an option in UAExpert to test this.
Thank you anyway!
